formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/ProtoBuf.kt

```diff
@@ -230,48 +235,84 @@ public class ProtoBuf(
         }
 
         fun writeDouble(value: Double, tag: Int) {
-            val header = encode32((tag shl 3) or i64)
-            val content = value.toLittleEndian().toByteArray()
-            out.write(header)
-            out.write(content)
+            out.encode32((tag shl 3) or i64)
+            out.writeLong(value.reverseBytes())
         }
 
         fun writeFloat(value: Float, tag: Int) {
-            val header = encode32((tag shl 3) or i32)
-            val content = value.toLittleEndian().toByteArray()
-            out.write(header)
-            out.write(content)
+            out.encode32((tag shl 3) or i32)
+            out.writeInt(value.reverseBytes())
         }
 
-        private fun encode32(number: Int, format: ProtoNumberType = ProtoNumberType.DEFAULT): ByteArray =
+        private fun Output.encode32(
+            number: Int,
+            format: ProtoNumberType = ProtoNumberType.DEFAULT
+        ) {
             when (format) {
-                ProtoNumberType.FIXED -> number.toLittleEndian().toByteArray()
-                ProtoNumberType.DEFAULT -> encodeVarint(number.toLong())
-                ProtoNumberType.SIGNED -> encodeVarint(((number shl 1) xor (number shr 31)))
+                ProtoNumberType.FIXED -> out.writeInt(number.reverseBytes())
+                ProtoNumberType.DEFAULT -> encodeVarint64(number.toLong())
+                ProtoNumberType.SIGNED -> encodeVarint32(((number shl 1) xor (number shr 31)))
             }
+        }
 
-
-        private fun encode64(number: Long, format: ProtoNumberType = ProtoNumberType.DEFAULT): ByteArray =
+        private fun Output.encode64(number: Long, format: ProtoNumberType = ProtoNumberType.DEFAULT) {
             when (format) {
-                ProtoNumberType.FIXED -> number.toLittleEndian().toByteArray()
-                ProtoNumberType.DEFAULT -> encodeVarint(number)
-                ProtoNumberType.SIGNED -> encodeVarint((number shl 1) xor (number shr 63))
+                ProtoNumberType.FIXED -> out.writeLong(number.reverseBytes())
+                ProtoNumberType.DEFAULT -> encodeVarint64(number)
+                ProtoNumberType.SIGNED -> encodeVarint64((number shl 1) xor (number shr 63))
             }
+        }
     }
 
-    private open inner class ProtobufReader(val decoder: ProtobufDecoder) : TaggedDecoder<ProtoDesc>() {
+    private open inner class ProtobufReader(@JvmField val decoder: ProtobufDecoder) : TaggedDecoder<ProtoDesc>() {
         override val context: SerialModule
             get() = this@ProtoBuf.context
 
-        private val indexByTag: MutableMap<Int, Int> = mutableMapOf()
-        private fun findIndexByTag(desc: SerialDescriptor, serialId: Int, zeroBasedDefault: Boolean = false): Int =
-            (0 until desc.elementsCount).firstOrNull {
-                extractParameters(
+        private var indexCache: IntArray? = null
```

Comment:
WDYT of using something like https://github.com/Kotlin/kotlinx.serialization/pull/597/files#diff-bcaa7635dcd4f1d128805c3c282f8253R17 ? 

Currently, it looks like that this cache would be built each time `ProtobufReader` is created, basically making it useless, e.g. in `List<X>` cache for X is built for every element, since in `beginStructure` ProtobufReader is re-created.

Reply:
Current cache approach is **still** better than the previous quadratic algorithm.
I'll fix this one in follow-up work, I am not ready to do one more performance analysis for collections right now.